framework/bundles/org.jboss.tools.rsp.server/src/main/java/org/jboss/tools/rsp/server/filewatcher/FileWatcherService.java

```diff
@@ -0,0 +1,584 @@
+		}
+	}
+
+	private void updateSubscriptionsForRemovedRegistration(Path path) {
+		// If I'm not needed, unsubscribe
+		if( !pathShouldBeSubscribed(path)) {
+			removeSubscription(path);
+			
+			// If I don't need to be watched anymore, do my parents? 
+			Path working = path.getParent();
+			while( working != null ) {
+				if( !pathShouldBeSubscribed(working)) {
+					removeSubscription(working);
+				} else {
+					break;
+				}
+				working = working.getParent();
+			}
+			
+		}
+		
+		/*
+		 *  I might still be needed, but maybe we don't need to watch
+		 *  every folder below me. For example, maybe someone subscribed
+		 *  to /home/user/test  (recursive), and another subscribed 
+		 *  to /home/user/test/inner/five.   
+		 *  
+		 *  We may still need to listen to changes in /home/user/test, 
+		 *  but perhaps nobody needs to hear updates from 
+		 *  /home/user/test/deployments or /home/user/test/incoming
+		 */
+		 
+		List<Path> nested = findAllSubtreeSubscriptions(path);
+		for( Path test : nested ) {
+			if( !pathShouldBeSubscribed(test)) {
+				removeSubscription(test);
+			}
+		}
+	}
+	
+	private List<Path> findAllSubtreeSubscriptions(Path path) {
+		List<Path> nested = new ArrayList<Path>();
+		for( Path test : subscriptions.keySet()) {
+			if( test.startsWith(path)) {
+				nested.add(test);
+			}
+		}
+		return nested;
+	}
+	
+	private void removeSubscription(Path path) {
+		// Stop watching the given folder. 
+		WatchKey wk = subscriptions.get(path);
+		if( wk != null )
+			wk.cancel();
+		subscriptions.remove(path);
+	}
+
+
+	private boolean pathShouldBeSubscribed(Path path) {
+		// A recursive request to me or my parent exists, so I'm still needed
+		if( pathOrParentHasRecursiveRequest(path))
+			return true;
+		
+		// A request still exists for this exact path, so I'm still needed
+		if( requests.get(path) != null && requests.get(path).size() > 0 )
+			return true;
+		
+		// A request for a subfolder (or lower) still exists, so I'm still needed
+		Set<Path> requestPaths = requests.keySet();
+		for( Path testRequest : requestPaths) {
+			if( testRequest.startsWith(path))
+				return true;
+		}
+
+		// Nobody needs me
+		return false;
+	}
+	
+	private boolean pathOrParentHasRecursiveRequest(Path p) {
+		Path working = p;
+		while(working != null ) {
+			if( singlePathHasRecursiveRequest(working)) {
+				return true;
+			}
+			working = working.getParent();
+		}
+		return false;
+	}
+	private boolean singlePathHasRecursiveRequest(Path p) {
+		List<RegistrationRequest> list = requests.get(p);
+		if( list == null ) 
+			return false;
+		Iterator<RegistrationRequest> rit = list.iterator();
+		RegistrationRequest r = null;
+		while(rit.hasNext()) {
+			r = rit.next();
+			if( r.isRecursive())
+				return true;
+		}
+		return false;
+	}
+
+	@Override
+	public void run() {
+		WatchKey key;
+		try {
+			while (watchService != null && (key = watchService.take()) != null) {
+				// Get the list and reset immedietely
```

Comment:
misspelled ```immedietely```.
Can you please explain in the comment why you reset it (I imagine that this somehow marks the event as consumed?)?